cmd/ipfs/daemon.go

```diff
@@ -292,12 +293,22 @@ func daemonFunc(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment
 			return fmt.Errorf("fs-repo requires migration")
 		}
 
-		migrationCfg, err := readMigrationConfig(cctx.ConfigRoot)
+		// Read Migration section of IPFS config
+		migrationCfg, err := migrations.ReadMigrationConfig(cctx.ConfigRoot)
 		if err != nil {
 			return err
 		}
 
-		fetcher, err = getMigrationFetcher(migrationCfg, &cctx.ConfigRoot)
+		// Define function to create IPFS fetcher
+		newIpfsFetcher := func(distPath string) migrations.Fetcher {
+			return ipfsfetcher.NewIpfsFetcher(distPath, 0, &cctx.ConfigRoot)
+		}
```

Comment:
This confused me a lot. `GetMigrationFetcher` is supposed to create the fetchers and takes care of creating the Http one internally. However, the IPFS one needs to be provided directly, and no less than inside a function.

I take that this is primarily to be able to mock it while initializing it with distPath or... ?   The GetMigrationFetcher function as it was (taking care of creating the fetchers as needed) makes more sense to what I would expect such function to do. It is not very useful to have a function that creates fetchers if I have to provide one of them anyways.

Reply:
Different applications (ipfs daemon, ipfs-update) have different implementations of an IPFS fetcher, so they need to provide their own implementation.  When constructing the Fetcher (which may produce a MultiFetcher) by calling `migrations.GetMigrationFetcher`, depending on the migration config only certain types of fetcher will be needed.  For example, if config said "only use HTTP" or "only use IPFS".

We do not want to supply an already-constructed IPFS fetcher (which may be very expensive if it needs to set up a temporary node). So instead, we supply a function to construct the implementation's IPFS fetcher, only if an IPFS fetcher is needed according to the config.

Reply:
Added [comment in code](https://github.com/ipfs/go-ipfs/pull/8153/files#diff-73163a91b837dff9f7108907361d53ed19b735ca59e5231eed6b49ade275cacfR304-R309) to explain